src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs

```diff
@@ -762,31 +790,282 @@ public void Dispose()
+                _sender = null;
+            }
+        }
+
+        // This event is triggered when the ping is completed, and passes along the eventargs so that we know
+        // if the ping was cancelled, or an exception was thrown.
+        private static void OnPingComplete(object sender, PingCompletedEventArgs e)
+        {
+            ((TestConnectionCommand)e.UserState)._pingCompleteArgs = e;
+            ((TestConnectionCommand)e.UserState)._pingComplete.Set();
+        }
+
+        /// <summary>
+        /// The class contains information about the source, the destination and ping results.
+        /// </summary>
+        public class PingStatus
+        {
+            /// <summary>
+            /// Initializes a new instance of the <see cref="PingStatus"/> class.
+            /// This constructor allows manually specifying the initial values for the cases where the PingReply
+            /// object may be missing some information, specifically in the instances where PingReply objects are
+            /// utilised to perform a traceroute.
+            /// </summary>
+            /// <param name="source">The source machine name or IP of the ping.</param>
+            /// <param name="destination">The destination machine name of the ping.</param>
+            /// <param name="reply">The response from the ping attempt.</param>
+            /// <param name="options">The PingOptions specified when the ping was sent.</param>
+            /// <param name="latency">The latency of the ping.</param>
+            /// <param name="bufferSize">The buffer size.</param>
+            /// <param name="pingNum">The sequence number in the sequence of pings to the hop point.</param>
+            internal PingStatus(
+                string source,
+                string destination,
+                PingReply reply,
+                PingOptions options,
+                long latency,
+                int bufferSize,
+                uint pingNum)
+                : this(source, destination, reply, pingNum)
+            {
+                _options = options;
+                _bufferSize = bufferSize;
+                _latency = latency;
+            }
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="PingStatus"/> class.
+            /// </summary>
+            /// <param name="source">The source machine name or IP of the ping.</param>
+            /// <param name="destination">The destination machine name of the ping.</param>
+            /// <param name="reply">The response from the ping attempt.</param>
+            /// <param name="pingNum">The sequence number of the ping in the sequence of pings to the target.</param>
+            internal PingStatus(string source, string destination, PingReply reply, uint pingNum)
+            {
+                Ping = pingNum;
+                Reply = reply;
+                Source = source;
+                Destination = destination ?? reply.Address.ToString();
```

Comment:
I think it should be in calling code. The side effect is not good design.
```suggestion
                Destination = destination;
```